subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java

```diff
@@ -0,0 +1,240 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.component.internal;
+
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.artifacts.ConfigurationPublications;
+import org.gradle.api.artifacts.PublishArtifact;
+import org.gradle.api.artifacts.type.ArtifactTypeDefinition;
+import org.gradle.api.attributes.Bundling;
+import org.gradle.api.attributes.Category;
+import org.gradle.api.attributes.Usage;
+import org.gradle.api.attributes.VerificationType;
+import org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact;
+import org.gradle.api.internal.plugins.DefaultArtifactPublicationSet;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.api.internal.tasks.JvmConstants;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.BasePlugin;
+import org.gradle.api.plugins.ExtensionContainer;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.plugins.PluginContainer;
+import org.gradle.api.plugins.internal.DefaultAdhocSoftwareComponent;
+import org.gradle.api.plugins.internal.JavaConfigurationVariantMapping;
+import org.gradle.api.plugins.internal.JvmPluginsHelper;
+import org.gradle.api.plugins.jvm.internal.JvmPluginServices;
+import org.gradle.api.provider.ProviderFactory;
+import org.gradle.api.publish.PublishingExtension;
+import org.gradle.api.publish.internal.versionmapping.VersionMappingStrategyInternal;
+import org.gradle.api.publish.ivy.IvyPublication;
+import org.gradle.api.publish.ivy.internal.publication.IvyPublicationInternal;
+import org.gradle.api.publish.maven.MavenPublication;
+import org.gradle.api.publish.maven.internal.publication.MavenPublicationInternal;
+import org.gradle.api.publish.plugins.PublishingPlugin;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskContainer;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.bundling.Jar;
+import org.gradle.api.tasks.testing.Test;
+import org.gradle.internal.reflect.Instantiator;
+import org.gradle.jvm.component.JvmSoftwareComponent;
+
+import javax.inject.Inject;
+
+/**
+ * Default implementation of {@link JvmSoftwareComponent}.
+ * <p>
+ * This component was written assuming it will only be instantiated once, and therefore it hard-codes the names of the
+ * domain objects it creates. However, future iterations of this component should allow it to be more generic.
+ */
+public class DefaultJvmSoftwareComponent extends DefaultAdhocSoftwareComponent implements JvmSoftwareComponent {
```

Comment:
🤔 I think the hierarchy here needs some refactoring so that it makes more sense.  I think there are a couple of problems:

1. The new `JvmSoftwareComponent` contains variants, and should extend `ComponentWithVariants`.
2. `JvmSoftwareComponent` is no longer Adhoc, so it and it's implementation shouldn't extend those Adhoc types, as the goal seems to be that it produces a fixed set of variants that shouldn't be modified.
3. `AdhocComponentWithVariants` is *NOT* a `ComponentWithVariants`, which is confusing, as it should be.
4. `SoftwareComponentInternal` mentions that _This will be replaced by {@link org.gradle.api.component.ComponentWithVariants} and other public APIs._

Suggested New Hierarchy:
```
SoftwareComponent <- ComponentWithVariants <- AdhocComponentWithVariants
                                           <- JvmSoftwareComponent
```
- Where `ComponentWithVariants` defines the `getUsages` method from `SoftwareComponentInternal`, making it now public API (how would this be meaningful for a component without variants anyway?) and allows for deleting the `SCI` interface completely.

```
AdhocComponentWithVariants <- DefaultAdhocSoftwareComponent
JvmSoftwareComponent <- DefaultJvmSoftwareComponent
```

This seems like it would be more correct to me.

Reply:
Oh, I noticed on a closer read of the associated Spec "For backwards compatibility with existing usages of the java component, DefaultJvmSoftwareComponent will implement AdhocComponentWithVariants by extending DefaultAdhocSoftwareComponent."

Perhaps just (temporarily) extending the Adhoc interface would be enough.

Reply:
Yeah. We will eventually not extend `DefaultAdhocSoftwareComponent` but instead implement `AdhocComponentWithVariants` ourself, and then update all the interface methods to have deprecation warnings. 

However, we may need to keep this around a little longer since otherwise we do not have the proper APIs for users to interface with.

Reply:
I think my concern is with Adhoc being in the name in the hierarchy, when this is anything but adhoc - it's a strongly defined model for JVM components.  But I understand there are bigger concerns.

Reply:
Yeah. Adhoc should not be in the final hierarchy. However, only the internal implementation implements this interface. Users explicitly need to cast the component to this type if they want to use it in the adhoc fashion. The public type, `JvmSoftwareComponent` (which has been subsequently removed from this PR) will not include any adhoc types in its hierarchy.

Reply:
In terms of implementing `ComponentWithVariants`, this situation is pretty complex. `ComponentWithVariants` is more like `ComponentWithChildren` or `ComponentsWithRemoteVariants`. Each variant (which is also a `SoftwareComponent`) advertised by a `ComponentWithVariants` is used as a RemoteVariant (AKA "available-at" variant) in the published GMM. This variant/component must ALSO be separately published in their own publication. This is to get around the limitation that a publication can only publish a single POM and it must accept a component for its `from` parameter. 

Any component which has several non-runtime/compile variants (with differing dependencies) must be published in the above way if they are to remain compatible with Maven (and maybe ivy?). This can get pretty complicated since each component must now be logically separated into several sub-components just because of a limitation of our publishing API. 

Today, @tresat and I spoke for some time today and came up with a relatively complete solution. The root of the problem is that Gradle does not know how to publish complex components (KMP for example). A standard `MavenPublication` will never be able to publish arbitrary components to a repository. The mapping from variants to POMs is arbitrary and component-specific. For this reason, we concluded that components themselves must be repository ecosystem aware. 

Future plans involve designing and creating `MavenAwareSoftwareComponent` and `IvyAwareSoftwareComponent` which each include a method for returning maven and ivy publication metadata, respectively. Components can then implement these interfaces to inform the publication infrastructure how to publish itself. In the Maven example, a component would be able to define the (one or more) POMs required to publish itself, and at which coordinates those POMs should be uploaded to. 

After these APIs are complete, we will deprecate `ComponentWithVariants`. Existing "remote variant" like functionality will theoretically be able to be derived from the maven or ivy metadata a component advertises. 

These APIs will take some time to design and implement though in the log run I expect this to solve a lot of problems in the publishing space, including the potential to solve scala-related publishing issues we've seen recently.